SQL: missing serialisation of OffsetTime values for TIME type support

`SELECT count(*) AS c FROM test_emp WHERE birth_date::time IN ('00:00:00Z'::TIME)`

will fail with:
```
        "root_cause": [
            {
                "type": "illegal_argument_exception",
                "reason": "can not write type [class java.time.OffsetTime]",
                "stack_trace": "org.elasticsearch.ElasticsearchException$1: can not write type [class java.time.OffsetTime]
	at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:644)
	at org.elasticsearch.action.search.SearchPhaseExecutionException.guessRootCauses(SearchPhaseExecutionException.java:161)
	at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:626)
	at org.elasticsearch.ElasticsearchException.generateFailureXContent(ElasticsearchException.java:572)
	at org.elasticsearch.rest.BytesRestResponse.build(BytesRestResponse.java:149)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:110)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:93)
	at org.elasticsearch.rest.action.RestActionListener.onFailure(RestActionListener.java:58)
	at org.elasticsearch.client.node.NodeClient.lambda$executeLocally$1(NodeClient.java:94)
	at org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:184)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:50)
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71)
	at org.elasticsearch.xpack.sql.execution.PlanExecutor.lambda$sql$1(PlanExecutor.java:101)
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71)
	at org.elasticsearch.xpack.sql.execution.search.Querier$BaseActionListener.onFailure(Querier.java:583)
	at org.elasticsearch.client.node.NodeClient.lambda$executeLocally$1(NodeClient.java:94)
	at org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:184)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:50)
	at org.elasticsearch.action.ActionListener$5.onFailure(ActionListener.java:258)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.raisePhaseFailure(AbstractSearchAsyncAction.java:584)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:558)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:324)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:593)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:400)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:258)
	at org.elasticsearch.action.search.SearchExecutionStatsCollector.onFailure(SearchExecutionStatsCollector.java:73)
	at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:59)
	at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:408)
	at org.elasticsearch.transport.TransportService$6.handleException(TransportService.java:581)
	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1121)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1230)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1204)
	at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:61)
	at org.elasticsearch.transport.TransportChannel.sendErrorResponse(TransportChannel.java:56)
	at org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:51)
	at org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:88)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:39)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:733)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.lang.IllegalArgumentException: can not write type [class java.time.OffsetTime]
	at org.elasticsearch.common.io.stream.StreamOutput.writeGenericValue(StreamOutput.java:866)
	at org.elasticsearch.common.io.stream.StreamOutput.lambda$static$12(StreamOutput.java:693)
	at org.elasticsearch.common.io.stream.StreamOutput.writeGenericValue(StreamOutput.java:864)
	at org.elasticsearch.common.io.stream.StreamOutput.lambda$static$14(StreamOutput.java:718)
	at org.elasticsearch.common.io.stream.StreamOutput.writeGenericValue(StreamOutput.java:864)
	at org.elasticsearch.common.io.stream.StreamOutput.writeMap(StreamOutput.java:533)
	at org.elasticsearch.script.Script.writeTo(Script.java:573)
	at org.elasticsearch.index.query.ScriptQueryBuilder.doWriteTo(ScriptQueryBuilder.java:68)
	at org.elasticsearch.index.query.AbstractQueryBuilder.writeTo(AbstractQueryBuilder.java:79)
	at org.elasticsearch.common.io.stream.StreamOutput.writeNamedWriteable(StreamOutput.java:1133)
	at org.elasticsearch.common.io.stream.StreamOutput.writeOptionalNamedWriteable(StreamOutput.java:1144)
	at org.elasticsearch.search.builder.SearchSourceBuilder.writeTo(SearchSourceBuilder.java:271)
	at org.elasticsearch.common.io.stream.StreamOutput.writeOptionalWriteable(StreamOutput.java:994)
	at org.elasticsearch.search.internal.ShardSearchRequest.innerWriteTo(ShardSearchRequest.java:254)
	at org.elasticsearch.search.internal.ShardSearchRequest.cacheKey(ShardSearchRequest.java:401)
	at org.elasticsearch.indices.IndicesService.loadIntoContext(IndicesService.java:1387)
	at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:370)
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:431)
	at org.elasticsearch.search.SearchService$2.lambda$onResponse$0(SearchService.java:401)
	at org.elasticsearch.action.ActionRunnable.lambda$supply$0(ActionRunnable.java:58)
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	... 6 more
"
            }
        ]
```

--- a/server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java
+++ b/server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java
@@ -61,7 +61,10 @@ import java.nio.file.FileSystemLoopException;
 import java.nio.file.NoSuchFileException;
 import java.nio.file.NotDirectoryException;
 import java.time.Instant;
+import java.time.LocalTime;
+import java.time.OffsetTime;
 import java.time.ZoneId;
+import java.time.ZoneOffset;
 import java.time.ZonedDateTime;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -749,6 +752,8 @@ public abstract class StreamInput extends InputStream {
                 return readCollection(StreamInput::readGenericValue, HashSet::new, Collections.emptySet());
             case 26:
                 return readBigInteger();
+            case 27:
+                return readOffsetTime();
             default:
                 throw new IOException("Can't read unknown type [" + type + "]");
         }
@@ -794,6 +799,11 @@ public abstract class StreamInput extends InputStream {
     private ZonedDateTime readZonedDateTime() throws IOException {
         final String timeZoneId = readString();
         return ZonedDateTime.ofInstant(Instant.ofEpochMilli(readLong()), ZoneId.of(timeZoneId));
+    }
+
+    private OffsetTime readOffsetTime() throws IOException {
+        final String zoneOffsetId = readString();
+        return OffsetTime.of(LocalTime.ofNanoOfDay(readLong()), ZoneOffset.of(zoneOffsetId));
     }
 
     private static final Object[] EMPTY_OBJECT_ARRAY = new Object[0];

--- a/server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java
+++ b/server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java
@@ -59,6 +59,7 @@ import java.nio.file.FileSystemLoopException;
 import java.nio.file.NoSuchFileException;
 import java.nio.file.NotDirectoryException;
 import java.time.Instant;
+import java.time.OffsetTime;
 import java.time.ZoneId;
 import java.time.ZonedDateTime;
 import java.util.Arrays;
@@ -828,6 +829,15 @@ public abstract class StreamOutput extends OutputStream {
                         o.writeByte((byte) 26);
                         o.writeString(v.toString());
                     }
+            ),
+            entry(
+                OffsetTime.class,
+                (o, v) -> {
+                    o.writeByte((byte) 27);
+                    final OffsetTime offsetTime = (OffsetTime) v;
+                    o.writeString(offsetTime.getOffset().getId());
+                    o.writeLong(offsetTime.toLocalTime().toNanoOfDay());
+                }
             ));
 
     private static Class<?> getGenericType(Object value) {

--- a/server/src/test/java/org/elasticsearch/common/io/stream/BytesStreamsTests.java
+++ b/server/src/test/java/org/elasticsearch/common/io/stream/BytesStreamsTests.java
@@ -36,6 +36,7 @@ import org.joda.time.DateTimeZone;
 
 import java.io.EOFException;
 import java.io.IOException;
+import java.time.OffsetTime;
 import java.time.ZoneId;
 import java.util.ArrayList;
 import java.util.Collections;
@@ -312,6 +313,8 @@ public class BytesStreamsTests extends ESTestCase {
         out.writeOptionalTimeZone(DateTimeZone.getDefault());
         out.writeOptionalTimeZone(null);
         out.writeGenericValue(new DateTime(123456, DateTimeZone.forID("America/Los_Angeles")));
+        final OffsetTime offsetNow = OffsetTime.now(randomZone());
+        out.writeGenericValue(offsetNow);
         final byte[] bytes = BytesReference.toBytes(out.bytes());
         StreamInput in = StreamInput.wrap(BytesReference.toBytes(out.bytes()));
         assertEquals(in.available(), bytes.length);
@@ -349,6 +352,7 @@ public class BytesStreamsTests extends ESTestCase {
         JodaCompatibleZonedDateTime jdt = (JodaCompatibleZonedDateTime) dt;
         assertThat(jdt.getZonedDateTime().toInstant().toEpochMilli(), equalTo(123456L));
         assertThat(jdt.getZonedDateTime().getZone(), equalTo(ZoneId.of("America/Los_Angeles")));
+        assertThat(in.readGenericValue(), equalTo(offsetNow));
         assertEquals(0, in.available());
         IllegalArgumentException ex = expectThrows(IllegalArgumentException.class, () -> out.writeGenericValue(new Object() {
             @Override

